{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "668fc162-fda6-4854-818f-45e529e5b272",
   "metadata": {
    "tags": []
   },
   "source": [
    "# QUEEN Cloning Script for a Large Catalog of Plasmids\n",
    "### Samuel King | Yachie Lab | Last updated: January 2023\n",
    "\n",
    "This is a QUEEN script for constructing plasmids by Digestion-Ligation, Gibson Assembly, and Golden Gate Assembly for the plasmids described in the CloneSelect paper."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d70f598-dbf8-46cc-8d1d-be1c1ab2ced7",
   "metadata": {},
   "source": [
    "### Setting up QUEEN\n",
    "\n",
    "* I recommend using Anaconda as a distribution and coding your QUEEN scripts in JupyterLab.\n",
    "* This tutorial https://biocircuits.github.io/chapters/00_setting_up_python_computing_environment.html describes very clearly how to set up your environment with Anaconda.\n",
    "* Write out the exact cloning plan for your desired plasmid step-by-step, including template plasmid names, primer names, enzyme names, etc., before starting your QUEEN script.\n",
    "* If designing many plasmids, a spreadsheet is useful to carry the information. Then, you can code your functions to work with the spreadsheet items.\n",
    "* The full tutorial for QUEEN can be found here: https://github.com/yachielab/QUEEN\n",
    "\n",
    "### After setting up Anaconda and getting JupyterLab going, start your script like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e106fa51-bb9a-4d7d-849f-bb848cdd2fad",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib agg\n",
    "\n",
    "import sys\n",
    "from QUEEN.queen import *\n",
    "set_namespace(globals())\n",
    "from QUEEN import cutsite as cs\n",
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "from Bio import GenBank, SeqIO\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "91aa5d8b-2a1a-49b8-9edb-26fd652ac2d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load primers, plasmids, and assemblies\n",
    "file_path = '../QUEEN/Assemblies Information Table_revise_SI1.xlsx'\n",
    "oligos = pd.read_excel(file_path, sheet_name=0)\n",
    "plasmids = pd.read_excel(file_path, sheet_name=1)\n",
    "digestion_ligation_assemblies = pd.read_excel(file_path, sheet_name=2)\n",
    "golden_gate_assemblies = pd.read_excel(file_path, sheet_name=3)\n",
    "gibson_assemblies = pd.read_excel(file_path, sheet_name=4)\n",
    "processes = pd.read_excel(file_path, sheet_name=5)\n",
    "\n",
    "#Generate oligo QUEEN object list\n",
    "oligo_list = []\n",
    "i = 0\n",
    "while i < len(oligos):\n",
    "    primer = QUEEN(seq = oligos['seq'][i], product = oligos['primer_id'][i])\n",
    "    oligo_list.append(primer)\n",
    "    i = i+1\n",
    "\n",
    "\n",
    "# this takes HTTP requests\n",
    "plasmid_list = []\n",
    "i = 0\n",
    "while i < len(plasmids):\n",
    "    plasmid = QUEEN(record = str(plasmids['record'][i]), dbtype = str(plasmids['dbtype'][i]), product = str(plasmids['plasmid_id'][i]))\n",
    "    plasmid_list.append(plasmid)\n",
    "    time.sleep(1)\n",
    "\n",
    "    i = i+1\n",
    "\n",
    "oligo_name_list = []\n",
    "i = 0\n",
    "while i < len(oligos['primer_id']):\n",
    "    name = oligos['primer_id'][i]\n",
    "    oligo_name_list.append(name)\n",
    "    i = i+1\n",
    "\n",
    "plasmid_name_list = []\n",
    "i = 0\n",
    "while i < len(plasmids['plasmid_id']):\n",
    "    name = plasmids['plasmid_id'][i]\n",
    "    plasmid_name_list.append(name)\n",
    "    i = i+1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cce42f81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pSI_356_v2', 'pSI-894', 'pLV-CS-282_v2', 'pLV-CS-281_v4', 'lentiGuide_Puro']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plasmid_list\n",
    "oligo_list \n",
    "plasmid_name_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6f5bb5b-662e-44c4-8c1a-eaa5cba8f647",
   "metadata": {},
   "source": [
    "### Golden_Gate_Assembly_2 Template\n",
    "gRNA and reporter construction with BsmBI Golden Gate Assembly\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ad3c415c-bfd5-4b34-975f-74e12f6b28ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set up the data for plasmids constructed by Golden_Gate_Assembly_2\n",
    "assembly_method = golden_gate_assemblies['assembly_method'] #extract the assembly_method column from the golden_gate sheet\n",
    "golden_gate_assembly_2 = ['Golden_Gate_Assembly_2'] #set up for extracting rows with Golden_Gate_Assembly_2\n",
    "golden_gate_assembly_2 = golden_gate_assemblies.query('@golden_gate_assembly_2 in assembly_method') #extract only rows containing Golden_Gate_Assembly_2\n",
    "golden_gate_assembly_2.index = range(len(golden_gate_assembly_2)) #re-index the extracted list to be in the order of only the rows containing Golden_Gate_Assembly_2\n",
    "\n",
    "def find_primer_binding_str(primer_type,  #'FW' or 'RV'\n",
    "                            primer,       #QUEEN object\n",
    "                            template      #QUEEN object\n",
    "                           ):\n",
    "    \"\"\"\n",
    "    Given a type of primer ('FW' or 'RV'), the primer, and template sequence, finds the initial binding site of the primer in the template sequence, whether the primer has an overhang or not.\n",
    "    Returns the string sequence of the portion of the primer that binds the template sequence.\n",
    "    Requires minimum 18bp annealing between primer and template. The 3' end of the binding primer must anneal.\n",
    "    \"\"\"\n",
    "    # Note: re.search cannot search in the 3' to 5' direction on a QUEEN DNA object, so the function will not work for Rv primers unless the template is flipped\n",
    "    # For example:\n",
    "      # match1 = re.search('hell', 'hello') #Finds match='hell'\n",
    "      # print(match1)\n",
    "      # match2 = re.search('lleh', 'hello') #Finds None\n",
    "      # print(match2)\n",
    "    \n",
    "    n = len(primer.seq)                  #length of primer\n",
    "    i=-18                                #starting point of i\n",
    "    flippedtemplate = flipdna(template)  #reverse complement of template\n",
    "    binding_site = []\n",
    "    \n",
    "    \n",
    "    if primer_type == 'FW':                                 #for searching templates with Fw primer\n",
    "        while i <= n:\n",
    "            match = re.search(primer.seq[i:], template.seq) #search for each character in the primer string until the maximum amount of the primer matches the template string\n",
    "            binding_site.append(match)                      #add matching string to the list binding_site\n",
    "            if match == None:                               #end loop when there are no more matches\n",
    "                break\n",
    "            elif -i == n:                                   #end loop when there are no is no more primer string to iterate on\n",
    "                break\n",
    "            i = i-1\n",
    "    elif primer_type == 'RV':                               #for searching templates with Rv primer; template must be reverse complement because re can't search on dsDNA like QUEEN can\n",
    "        while i <= n:\n",
    "            match = re.search(primer.seq[i:], flippedtemplate.seq)\n",
    "            binding_site.append(match)\n",
    "            if match == None:\n",
    "                break\n",
    "            elif -i == n:\n",
    "                break\n",
    "            i = i-1\n",
    "            \n",
    "    if len(binding_site) == 1 and binding_site[-1] == None:  #if there is no binding sequence, the binding_site list will only contain 1 NoneType object\n",
    "        return 'No primer binding site found.'               #notify that no binding sequence for the primer was found in the template\n",
    "    elif binding_site[-1] == None:                           #the while loop ends the binding_site list with a NoneType object if the primer has an overhang\n",
    "        return binding_site[-2].group()                      #show the last matching binding site for a primer with an overhang\n",
    "    else:\n",
    "        return binding_site[-1].group()                      #show the last matching binding site for a primer with no overhang\n",
    "        \n",
    "\n",
    "def find_primer_binding_primer_str(fw_primer,  #QUEEN object\n",
    "                                   rv_primer   #QUEEN object\n",
    "                                  ):\n",
    "    \"\"\"\n",
    "    Searches for binding sites between two primers (useful for template-free PCR or annealing of two ssDNA oligos).\n",
    "    Returns string of the binding portion of the primer that anneals to the other primer. 3' end of primer must anneal.\n",
    "    If primers anneal and are equal size, returns string of the binding portion of the primer provided second.\n",
    "    Requires minimum 18-bp annealing between two oligos.\n",
    "    \"\"\"\n",
    "    \n",
    "    #The 'RV' feature must be set for every search because by design, every template (whether it's a Fw primer or Rv primer),\n",
    "    #will be antisense to the query, unlike when we're searching on a plasmid, which has a sense strand and an antisense strand\n",
    "    \n",
    "    if len(fw_primer.seq) > len(rv_primer.seq):                         #The smaller primer needs to be the query and the larger primer needs to be the template searched on.\n",
    "        return find_primer_binding_str('RV', rv_primer, fw_primer)      \n",
    "    elif len(fw_primer.seq) < len(rv_primer.seq):\n",
    "        return find_primer_binding_str('RV', fw_primer, rv_primer)\n",
    "    elif len(fw_primer.seq) == len(rv_primer.seq):\n",
    "        return find_primer_binding_str('RV', fw_primer, rv_primer)\n",
    "\n",
    "\n",
    "def typeIIS_digest_backbone(dna,  #QUEEN object\n",
    "                            re,   #str\n",
    "                            pn,   #str\n",
    "                            pd    #str\n",
    "                           ):\n",
    "    \"\"\"\n",
    "    Given a QUEEN DNA sequence for a backbone, finds the Type II-S restriction site(s) and digests at the found sites.\n",
    "    User may specify a process name (pn) and process description (pd) as strings.\n",
    "    Returns the cropdna backbone object. Also provides length and the object's topology (linear in this case).\n",
    "    \"\"\"\n",
    "    from QUEEN import cutsite #Import a restriction enzyme library \n",
    "    \n",
    "    sites = dna.searchsequence(cutsite.lib[re])\n",
    "    fragments = cropdna(dna, *sites, pn=pn, pd=pd) #Produces the backbone fragment\n",
    "    return fragments #Keep the backbone fragment\n",
    "\n",
    "def anneal_oligos_object(fw_primer,  #QUEEN object\n",
    "                         rv_primer,  #QUEEN object\n",
    "                         pn,         #str\n",
    "                         pd          #str\n",
    "                        ):\n",
    "    \"\"\"\n",
    "    Given Fw and Rv primers, creates a new fragment through ssDNA annealing of the two primers to each other.\n",
    "    *Requirement: Primers must anneal with at least 18 bp of binding.\n",
    "    Leaves overhangs on either, neither, or both sides depending on the design of the two oligos.\n",
    "    User may specify a process name (pn) and process description (pd) as strings.\n",
    "    Returns the QUEEN object with name \" \". Also provides length and the object's topology (linear in this case).\n",
    "    \"\"\"\n",
    "    #\n",
    "    # To build dsDNAs from two ssDNA oligos annealing (but not extending), we must consider the nine possible combinations of oligos:\n",
    "    #\n",
    "    #               Case 1           Case 2                           Case 3                           Case 4                           Case 5\n",
    "    # Fw primer:    5'-anneal-3'     5'-overhang-anneal-overhang-3'   5'----------anneal----------3'   5'-overhang-anneal----------3'   5'----------anneal-overhang-3'\n",
    "    # Rv primer:    3'-anneal-5'     3'----------anneal----------5'   3'-overhang-anneal-overhang-5'   3'----------anneal-overhang-5'   3'-overhang-anneal----------5'\n",
    "    #                  Frag.1            Frag.2          Frag.3           Frag.4          Frag.5   \n",
    "    #\n",
    "    #                                Case 6                           Case 7                           Case 8                           Case 9\n",
    "    # Fw primer:                     5'-overhang-anneal-3'            5'----------anneal-3'            5'-anneal-overhang-3'            5'-anneal----------3'\n",
    "    # Rv primer:                     3'----------anneal-5'            3'-overhang-anneal-5'            3'-anneal----------5'            3'-anneal-overhang-5'\n",
    "    #\n",
    "    #\n",
    "    # With the information of two primer sequences we must generate 5 unique fragments that will exist or not exist depending on which IF statement is satisified.\n",
    "    # Then, we will concatenate fragments depending on the condition satisfied.\n",
    "    #\n",
    "    # Approach: Generate each fragment, then concatenate them on the ds_anneal string, and the product should be one of cases 1-9.\n",
    "    # If an annealed oligo doesn't exist, it should notify the user.\n",
    "    #\n",
    "    \n",
    "    fw_primer_str = fw_primer.seq                                            #String of Fw primer sequence\n",
    "    rv_primer_str = rv_primer.seq                                            #String of Rv primer sequence\n",
    "    \n",
    "    if fw_primer_str == rv_primer_str:                                       #Function can't run if oligos are exactly identical\n",
    "        return 'Identical ssDNAs cannot anneal. Please revise your entry.'\n",
    "    elif len(fw_primer_str) < 18:                                            #Function can't run if oligos have <18 bp annealing.\n",
    "        return 'Oligos could not anneal because at least one is too short. Please revise your entry.'\n",
    "    elif len(rv_primer_str) < 18:                                            #Function can't run if oligos have <18 bp annealing.\n",
    "        return 'Oligos could not anneal because at least one is too short. Please revise your entry.'\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "    fw_overhang_list = []                                                    #Empty list for Fw overhang(s)\n",
    "    rv_overhang_list = []                                                    #Empty list for Rv overhang(s)\n",
    "    fw_fragment_list = []                                                    #Empty list for Fw overhang fragment(s)\n",
    "    rv_fragment_list = []                                                    #Empty list for Fw overhang fragment(s)\n",
    "    \n",
    "    #Generate Frag.1\n",
    "    anneal = find_primer_binding_primer_str(fw_primer, rv_primer)            #Finds the annealing site between the two primers\n",
    "    if fw_primer_str.__contains__(anneal) == True:\n",
    "        ds_anneal = generate_dsdna(anneal)                                   #Generates the double-stranded sequence of the binding sites of the two primers upon annealing\n",
    "    elif fw_primer_str.__contains__(anneal) == False:                        #Error for Case 5 occurs here because the 3' end of the Rv primer doesn't anneal\n",
    "        ds_anneal = generate_dsdna(flipdna(anneal))                          #Generates the double-stranded sequence of the binding sites of the two primers upon annealing\n",
    "    \n",
    "    \n",
    "    #Generate Frag.2 and Frag.3\n",
    "    overhangs_fw_primer = primer_overhangs(fw_primer, anneal)\n",
    "    five_fw = five_prime_overhang(overhangs_fw_primer)                       #Provides the 5' overhang of the Fw primer (if any)\n",
    "    fw_overhang_list.append(five_fw)\n",
    "    three_fw = three_prime_overhang(overhangs_fw_primer)                     #Provides the 3' overhang of the Fw primer (if any)\n",
    "    fw_overhang_list.append(three_fw)\n",
    "    \n",
    "    #Position 0 of fw_fragment_list will correspond to Frag.2\n",
    "    #Position 1 of fw_fragment_list will correspond to Frag.3\n",
    "    i=0\n",
    "    while i < 2:\n",
    "        if fw_overhang_list[i] != '':\n",
    "            fragment = QUEEN(seq=fw_overhang_list[i] + '/' + ('-' * len(fw_overhang_list[i])))\n",
    "            fw_fragment_list.append(fragment)\n",
    "        elif fw_overhang_list[i] == '':\n",
    "            fw_fragment_list.append('No overhang')\n",
    "        elif i == 2:\n",
    "            break\n",
    "        i = i+1\n",
    "    \n",
    "    \n",
    "    #Generate Frag.4 and Frag.5\n",
    "    overhangs_rv_primer = primer_overhangs(rv_primer, anneal)\n",
    "    five_rv = five_prime_overhang(overhangs_rv_primer)                       #Provides the 5' overhang of the Rv primer (if any)\n",
    "    rv_overhang_list.append(five_rv)\n",
    "    three_rv = three_prime_overhang(overhangs_rv_primer)                     #Provides the 3' overhang of the Rv primer (if any)\n",
    "    rv_overhang_list.append(three_rv)\n",
    "    \n",
    "    #Position 0 of rv_fragment_list will correspond to Frag.4\n",
    "    #Position 1 of rv_fragment_list will correspond to Frag.5\n",
    "    i=0\n",
    "    while i < 2:\n",
    "        if rv_overhang_list[i] != '':\n",
    "            fragment = QUEEN(seq=rv_overhang_list[i] + '/' + ('-' * len(rv_overhang_list[i])))\n",
    "            rv_fragment_list.append(fragment)\n",
    "        elif rv_overhang_list[i] == '':\n",
    "            rv_fragment_list.append('No overhang')\n",
    "        elif i == 2:\n",
    "            break\n",
    "        i = i+1\n",
    "\n",
    "    \n",
    "    if anneal != 'No primer binding site found.':\n",
    "        \n",
    "        if fw_fragment_list[0] != 'No overhang':    #Add upstream Fw overhang if possible\n",
    "            frag2 = QUEEN(seq=fw_fragment_list[0].seq + ds_anneal.seq[0] + '/' + '-' * len(fw_fragment_list[0].seq) + generate_complement(ds_anneal.seq[0]))\n",
    "            ds_anneal_left_trim = cropdna(ds_anneal, 1, len(ds_anneal.seq))\n",
    "            anneal_left = joindna(frag2, ds_anneal_left_trim, pn=pn, pd=pd)\n",
    "            \n",
    "        elif rv_fragment_list[1] != 'No overhang':  #Add upstream Rv overhang if possible\n",
    "            frag4 = QUEEN(seq=('-' * len(rv_fragment_list[1].seq) + ds_anneal.seq[0] + '/' + rv_fragment_list[1].seq[::-1] + generate_complement(ds_anneal.seq[0])))\n",
    "            ds_anneal_left_trim = cropdna(ds_anneal, 1, len(ds_anneal.seq))\n",
    "            anneal_left = joindna(frag4, ds_anneal_left_trim, pn=pn, pd=pd)\n",
    "            \n",
    "        else:\n",
    "            anneal_left = ds_anneal\n",
    "    \n",
    "        if fw_fragment_list[1] != 'No overhang':    #Add downstream Fw overhang if possible\n",
    "            frag3 = QUEEN(seq=ds_anneal.seq[-1] + fw_fragment_list[1].seq + '/' + generate_complement(ds_anneal.seq[-1]) + ('-' * len(fw_fragment_list[1].seq)))\n",
    "            ds_anneal_right_trim = cropdna(anneal_left, 0, (len(anneal_left.seq) - 1))\n",
    "            annealed_oligo = joindna(ds_anneal_right_trim, frag3, pn=pn, pd=pd)\n",
    "            return annealed_oligo\n",
    "            \n",
    "        elif rv_fragment_list[0] != 'No overhang':  #Add downstream Rv overhang if possible\n",
    "            frag5 = QUEEN(seq= ds_anneal.seq[-1] + ('-' * len(rv_fragment_list[0].seq)) + '/' + generate_complement(ds_anneal.seq[-1]) + rv_fragment_list[0].seq[::-1])\n",
    "            ds_anneal_right_trim = cropdna(anneal_left, 0, (len(anneal_left.seq) - 1))\n",
    "            annealed_oligo = joindna(ds_anneal_right_trim, frag5, pn=pn, pd=pd)\n",
    "            return annealed_oligo\n",
    "        \n",
    "        elif fw_fragment_list[0] != 'No overhang':\n",
    "            return anneal_left\n",
    "            \n",
    "        elif rv_fragment_list[1] != 'No overhang':\n",
    "            return anneal_left\n",
    "        \n",
    "        else:\n",
    "            annealed_oligo = ds_anneal\n",
    "            return annealed_oligo\n",
    "            \n",
    "    else:\n",
    "        'Oligos could not anneal.'\n",
    "    \n",
    "\n",
    "def generate_dsdna(dna  #str\n",
    "                  ):\n",
    "    \"\"\"\n",
    "    Given a single-stranded string of DNA in UPPERCASE letters only, returns the double-stranded QUEEN object.\n",
    "    \"\"\"\n",
    "    sense = dna                                  #sense strand of dna\n",
    "    antisense = generate_complement(dna)         #antisense strand of dna\n",
    "    concatenate = sense + '/' + antisense        #concatenate into one string for QUEEN to recognize the object\n",
    "    dsdna = QUEEN(seq=concatenate)               #generate dsDNA\n",
    "    return dsdna                                 #print dsDNA\n",
    "\n",
    "\n",
    "def generate_complement(dna #str\n",
    "                       ):\n",
    "    \"\"\"\n",
    "    Given a string of DNA in UPPERCASE letters only, returns complement of the DNA (not the reverse complement).\n",
    "    \"\"\"\n",
    "    complementary_dna = dna.replace('A', 't').replace('T', 'a').replace('G', 'c').replace('C', 'g')\n",
    "    return complementary_dna\n",
    "\n",
    "\n",
    "def primer_overhangs(primer, #QUEEN object\n",
    "                     anneal  #str\n",
    "                    ):\n",
    "    \"\"\"\n",
    "    Given a primer and its known annealing sequence to another primer (*required to anneal*), determines the overhangs on its 5' and 3' ends (if any).\n",
    "    Returns the overhangs in the format 'NNN|NNN', where '|' separates the 5' overhang (left of the | ) and the 3' overhang (right of the bar | ).\n",
    "    There can be 0 or more N's on either side of the '|'.\n",
    "    Helper function for anneal_oligos.\n",
    "    \"\"\"\n",
    "    p = primer.seq\n",
    "    \n",
    "    if p.__contains__(anneal) == True:\n",
    "        return p.replace(anneal, '|')\n",
    "    elif p.__contains__(anneal) == False:\n",
    "        return p.replace(flipdna(anneal), '|')\n",
    "    \n",
    "def five_prime_overhang(string,\n",
    "                       ):\n",
    "    \"\"\"\n",
    "    Given a string of format 'NNN|NNN' where N can represent any nucleotide or absence of a nucleotide, returns the 5' side of the string (characters before the '|').\n",
    "    Helper function for anneal_oligos.\n",
    "    \"\"\" \n",
    "    i = 0\n",
    "    string_new = ''\n",
    "    n = len(string)\n",
    "    \n",
    "    while i <= n:\n",
    "        string_new = string_new + string[i]\n",
    "        if string[i] == '|':\n",
    "            break\n",
    "        elif i == n:\n",
    "            break\n",
    "        i = i+1\n",
    "        \n",
    "    return string_new[:i]\n",
    "\n",
    "def three_prime_overhang(string,\n",
    "                        ):\n",
    "    \"\"\"\n",
    "    Given a string of format 'NNN|NNN' where N can represent any nucleotide or absence of a nucleotide, returns the 3' side of the string (characters after the '|').\n",
    "    Helper function for anneal_oligos.\n",
    "    \"\"\" \n",
    "    i = -1\n",
    "    string_new = ''\n",
    "    n = len(string)\n",
    "    \n",
    "    while i <= n:\n",
    "        string_new = string[i] + string_new\n",
    "        if string[i] == '|':\n",
    "            break\n",
    "        elif i == n:\n",
    "            break\n",
    "        i = i-1\n",
    "    \n",
    "    j = i+1 \n",
    "    if string_new == '|':   #When '|' is the first character that i sees, then string_new becomes '|' and the loop ends. We remove this to give '' in such cases.\n",
    "        return ''\n",
    "    else:\n",
    "        return string_new[j:]\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4b48525",
   "metadata": {},
   "source": [
    "## Golden Gate Assembly Process\n",
    "For gRNA and CRISPRa reporters, used for revised experiments\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7122203b-56a8-4b24-8e84-102ddb4105b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set assembly steps\n",
    "\n",
    "#Step 1\n",
    "pn1 = processes['pn1'][3]\n",
    "pd1 = processes['pd1'][3]\n",
    "#Step 2\n",
    "pn2 = processes['pn2'][3]\n",
    "pd2 = processes['pd2'][3]\n",
    "#Step 3\n",
    "pn3 = processes['pn3'][3]\n",
    "pd3 = processes['pd3'][3]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "48b7be4d-0b2c-495d-81bf-fe4b394a9f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step 1: Digest backbone\n",
    "\n",
    "digested_backbones = []\n",
    "i = 0\n",
    "while i < len(golden_gate_assembly_2):\n",
    "    index_backbone = plasmid_name_list.index(golden_gate_assembly_2['backbone_id'][i])\n",
    "    re1 = golden_gate_assembly_2['re1'][i]\n",
    "    digested_backbone = typeIIS_digest_backbone(plasmid_list[index_backbone], re1, pn1, pd1)\n",
    "    digested_backbones.append(digested_backbone)\n",
    "    i = i+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c4ebabe-24c7-4a92-94f2-f2e2d7345d2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<queen.QUEEN object; project='dna_102', length='29 bp', sequence='CACCGCCTTATGACCCTGACACGCTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_106', length='29 bp', sequence='CACCGATTTAACTGGGCACGTTTGGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_110', length='28 bp', sequence='CACCGGTTACCTCCACGTCGAACGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_114', length='29 bp', sequence='CACCGAGCGTGTCAGGGTGACCGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_118', length='29 bp', sequence='CACCGAGTCTGTCTCTCACAGCGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_122', length='29 bp', sequence='CACCGAGTCTGGCAGTCACTGGGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_126', length='29 bp', sequence='CACCGTGCGTGTCTCTGTCTCGGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_130', length='29 bp', sequence='CACCGAGTCACGGAGCCAGTCGGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_134', length='29 bp', sequence='CACCCTCTCAGGCTCTGACTCTGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_138', length='29 bp', sequence='CACCGAGCGTGTCAGGGTCATAAGGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_142', length='29 bp', sequence='CACCGCCAAACGTGCCCAGTTAAATGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_146', length='29 bp', sequence='CACCGCGTTCGACGTGGAGGTAACCGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_150', length='29 bp', sequence='CACCGAGCGTGTCAGGGTGACCGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_154', length='29 bp', sequence='CACCGAGTCTGTCTCTCACAGCGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_158', length='29 bp', sequence='CACCGAGTCTGGCAGTCACTGGGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_162', length='29 bp', sequence='CACCGTGCGTGTCTCTGTCTCGGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_168', length='29 bp', sequence='CACCGAGTCACGGAGCCAGTCGGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_178', length='29 bp', sequence='CACCCTCTCAGGCTCTGACTCTGTGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_188', length='29 bp', sequence='CACCGAGCGTGTCAGGGTCATAAGGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_198', length='29 bp', sequence='CACCGCCAAACGTGCCCAGTTAAATGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_208', length='29 bp', sequence='CACCGCGTTCGACGTGGAGGTAACCGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_218', length='29 bp', sequence='CACCGCACGGTCACCCTGACACGCTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_228', length='29 bp', sequence='CACCGCACGCTGTGAGAGACAGACTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_238', length='29 bp', sequence='CACCGCACCCAGTGACTGCCAGACTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_248', length='29 bp', sequence='CACCGCACCGAGACAGAGACACGCAGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_258', length='29 bp', sequence='CACCGCACCGACTGGCTCCGTGACTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_268', length='29 bp', sequence='CACCGCACAGAGTCAGAGCCTGAGAGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_278', length='29 bp', sequence='CACCGCCTTATGACCCTGACACGCTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_288', length='29 bp', sequence='CACCGATTTAACTGGGCACGTTTGGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_298', length='28 bp', sequence='CACCGGTTACCTCCACGTCGAACGGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_308', length='29 bp', sequence='CACCGTAAAACGTGTAGTTGGGAGTGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_318', length='28 bp', sequence='CACCGGGTCGACGGGTAAGATGATGTTT', topology='linear'>,\n",
       " <queen.QUEEN object; project='dna_328', length='29 bp', sequence='CACCGTCAAATTCGGTGAGGGGGATGTTT', topology='linear'>]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Step 2: Anneal ssDNAs\n",
    "\n",
    "annealed_DNA = []\n",
    "i = 0\n",
    "while i < len(golden_gate_assembly_2):\n",
    "    index_fw = oligo_name_list.index(golden_gate_assembly_2['fw_primer'][i])\n",
    "    index_rv = oligo_name_list.index(golden_gate_assembly_2['rv_primer'][i])\n",
    "    insert = anneal_oligos_object(oligo_list[index_fw],\n",
    "                                  oligo_list[index_rv],\n",
    "                                  pn2, pd2)\n",
    "    annealed_DNA.append(insert)\n",
    "    i = i+1\n",
    "\n",
    "annealed_DNA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "822465bf-b55f-49ad-851d-ccfa5037b83d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       pSI_827                 2975 bp    DNA     circular UNK 29-MAY-2024\n",
      "DEFINITION  .\n",
      "ACCESSION   pSI_827\n",
      "VERSION     pSI_827\n",
      "KEYWORDS    .\n",
      "SOURCE      .\n",
      "  ORGANISM  .\n",
      "            .\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..1764\n",
      "                     /organism=\"synthetic DNA construct\"\n",
      "                     /mol_type=\"other DNA\"\n",
      "                     /broken_feature=\"pSI_356_v2:0:4835:-:0..4835:1..1764\"\n",
      "     CDS             complement(462..1322)\n",
      "                     /codon_start=1\n",
      "                     /gene=\"bla\"\n",
      "                     /product=\"beta-lactamase\"\n",
      "                     /label=\"AmpR\"\n",
      "                     /note=\"confers resistance to ampicillin, carbenicillin, and\n",
      "                     related antibiotics\"\n",
      "                     /translation=\"MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYI\n",
      "                     ELDLNSGKILESFRPEERFPMMSTFKVLLCGAVLSRIDAGQEQLGRRIHYSQNDLVEYS\n",
      "                     PVTEKHLTDGMTVRELCSAAITMSDNTAANLLLTTIGGPKELTAFLHNMGDHVTRLDRW\n",
      "                     EPELNEAIPNDERDTTMPVAMATTLRKLLTGELLTLASRQQLIDWMEADKVAGPLLRSA\n",
      "                     LPAGWFIADKSGAGERGSRGIIAALGPDGKPSRIVVIYTTGSQATMDERNRQIAEIGAS\n",
      "                     LIKHW\"\n",
      "     primer_bind     1085..1104\n",
      "                     /label=\"Amp-R\"\n",
      "                     /note=\"Ampicillin resistance gene, reverse primer\"\n",
      "     promoter        complement(1323..1427)\n",
      "                     /gene=\"bla\"\n",
      "                     /label=\"AmpR promoter\"\n",
      "     primer_bind     1495..1513\n",
      "                     /label=\"pBRforEco\"\n",
      "                     /note=\"pBR322 vectors, upsteam of EcoRI site, forward\n",
      "                     primer\"\n",
      "     primer_bind     1687..1706\n",
      "                     /label=\"LKO.1 5'\"\n",
      "                     /note=\"Human U6 promoter, forward primer\"\n",
      "     misc_RNA        1786..1861\n",
      "                     /label=\"gRNA scaffold\"\n",
      "                     /note=\"guide RNA scaffold for the Streptococcus pyogenes\n",
      "                     CRISPR/Cas9 system\"\n",
      "     source          1786..2975\n",
      "                     /organism=\"synthetic DNA construct\"\n",
      "                     /mol_type=\"other DNA\"\n",
      "                     /broken_feature=\"pSI_356_v2:0:4835:-:0..4835:3646..4835\"\n",
      "     CDS             complement(1957..2280)\n",
      "                     /codon_start=1\n",
      "                     /gene=\"lacZ fragment\"\n",
      "                     /product=\"LacZ-alpha fragment of beta-galactosidase\"\n",
      "                     /label=\"lacZ-alpha\"\n",
      "                     /translation=\"MTMITPSLHACRSTLEDPRVPSSNSLAVVLQRRDWENPGVTQLNR\n",
      "                     LAAHPPFASWRNSEEARTDRPSQQLRSLNGEWRLMRYFLLTHLCGISHRIWCTLSTICS\n",
      "                     DAA\"\n",
      "     primer_bind     1962..1981\n",
      "                     /label=\"pRS-marker\"\n",
      "                     /note=\"pRS vectors, use to sequence yeast selectable\n",
      "                     marker\"\n",
      "     primer_bind     2175..2197\n",
      "                     /label=\"M13/pUC Forward\"\n",
      "                     /note=\"In lacZ gene\"\n",
      "     primer_bind     2189..2206\n",
      "                     /label=\"M13 Forward\"\n",
      "                     /note=\"In lacZ gene. Also called M13-F20 or M13 (-21)\n",
      "                     Forward\"\n",
      "     primer_bind     2190..2206\n",
      "                     /label=\"M13 fwd\"\n",
      "                     /note=\"common sequencing primer, one of multiple similar\n",
      "                     variants\"\n",
      "     misc_feature    2207..2263\n",
      "                     /label=\"MCS\"\n",
      "                     /note=\"pUC18/19 multiple cloning site\"\n",
      "     primer_bind     complement(2276..2292)\n",
      "                     /label=\"M13 rev\"\n",
      "                     /note=\"common sequencing primer, one of multiple similar\n",
      "                     variants\"\n",
      "     primer_bind     complement(2276..2292)\n",
      "                     /label=\"M13 Reverse\"\n",
      "                     /note=\"In lacZ gene. Also called M13-rev\"\n",
      "     primer_bind     complement(2289..2311)\n",
      "                     /label=\"M13/pUC Reverse\"\n",
      "                     /note=\"In lacZ gene\"\n",
      "     protein_bind    2300..2316\n",
      "                     /label=\"lac operator\"\n",
      "                     /bound_moiety=\"lac repressor encoded by lacI\"\n",
      "                     /note=\"The lac repressor binds to the lac operator to\n",
      "                     inhibit transcription in E. coli. This inhibition can be\n",
      "                     relieved by adding lactose or\n",
      "                     isopropyl-beta-D-thiogalactopyranoside (IPTG).\"\n",
      "     promoter        complement(2324..2354)\n",
      "                     /label=\"lac promoter\"\n",
      "                     /note=\"promoter for the E. coli lac operon\"\n",
      "     protein_bind    2369..2390\n",
      "                     /label=\"CAP binding site\"\n",
      "                     /bound_moiety=\"E. coli catabolite activator protein\"\n",
      "                     /note=\"CAP binding activates transcription in the presence\n",
      "                     of cAMP.\"\n",
      "     primer_bind     complement(2507..2524)\n",
      "                     /label=\"L4440\"\n",
      "                     /note=\"L4440 vector, forward primer\"\n",
      "     rep_origin      complement(join(2678..2975,1..291))\n",
      "                     /direction=LEFT\n",
      "                     /label=\"ori\"\n",
      "                     /note=\"high-copy-number ColE1/pMB1/pBR322/pUC origin of\n",
      "                     replication\"\n",
      "     primer_bind     complement(2758..2777)\n",
      "                     /label=\"pBR322ori-F\"\n",
      "                     /note=\"pBR322 origin, forward primer\"\n",
      "     source          1..2975\n",
      "                     /building_history_540=\"QUEEN.dna_dict['dna_3']–=–QUEEN(seq=\n",
      "                     'CACCG/----c')\"\n",
      "                     /building_history_540=\"\"\n",
      "                     /building_history_540=\"pSI_827-DBRALKZ5GODW3554GCGUZA6W\"\n",
      "                     /building_history_537=\"QUEEN.dna_dict['dna_0']–=–QUEEN(seq=\n",
      "                     'GCCTTATGACCCTGACACGCT/cggaatactgggactgtgcga')\"\n",
      "                     /building_history_537=\"\"\n",
      "                     /building_history_537=\"pSI_827-1D2GO47OYV3DAWFDP9B5EFIU\"\n",
      "                     /building_history_541=\"QUEEN.dna_dict['dna_99']–=–cropdna(Q\n",
      "                     UEEN.dna_dict['dna_0'],–start='1/1',–end='21/21')\"\n",
      "                     /building_history_541=\"start: 1; end: 21\"\n",
      "                     /building_history_541=\"pSI_827-8CHT6VYYIG1D9H0ZAPMEEFY9\"\n",
      "                     /building_history_542=\"QUEEN.dna_dict['dna_100']–=–joindna(\n",
      "                     *[QUEEN.dna_dict['dna_3'],–QUEEN.dna_dict['dna_99']],–topol\n",
      "                     ogy='linear',–process_name='Anneal–ssDNAs',–process_descrip\n",
      "                     tion='Insert–fragments–were–generated–by–annealing–two–sing\n",
      "                     le-stranded–DNAs–(see–Table–S2–for–details)–and–phosphoryla\n",
      "                     ted–with–T4–Polynucleotide–Kinase–(NEB–#M0201S–or–Takara–#2\n",
      "                     021S).')\"\n",
      "                     /building_history_542=\"topology: linear\"\n",
      "                     /building_history_542=\"pSI_827-DM65T6LPMVD9ADNA1H7TH9BG\"\n",
      "                     /building_history_544=\"QUEEN.dna_dict['dna_101']–=–cropdna(\n",
      "                     QUEEN.dna_dict['dna_100'],–start='0/0',–end='24/24')\"\n",
      "                     /building_history_544=\"start: 0; end: 24\"\n",
      "                     /building_history_544=\"pSI_827-B1KHI9QVCT2E8C5W51BXSPVU\"\n",
      "                     /building_history_543=\"QUEEN.dna_dict['dna_4']–=–QUEEN(seq=\n",
      "                     'T----/aCAAA')\"\n",
      "                     /building_history_543=\"\"\n",
      "                     /building_history_543=\"pSI_827-8CQ8F0797WGJEIU6W4Q0O67Q\"\n",
      "                     /building_history_545=\"QUEEN.dna_dict['dna_102']–=–joindna(\n",
      "                     *[QUEEN.dna_dict['dna_101'],–QUEEN.dna_dict['dna_4']],–topo\n",
      "                     logy='linear',–process_name='Anneal–ssDNAs',–process_descri\n",
      "                     ption='Insert–fragments–were–generated–by–annealing–two–sin\n",
      "                     gle-stranded–DNAs–(see–Table–S2–for–details)–and–phosphoryl\n",
      "                     ated–with–T4–Polynucleotide–Kinase–(NEB–#M0201S–or–Takara–#\n",
      "                     2021S).')\"\n",
      "                     /building_history_545=\"topology: linear\"\n",
      "                     /building_history_545=\"pSI_827-AFDAD4XIVKIXADNA1H7TH9BG\"\n",
      "                     /building_history_465=\"QUEEN.dna_dict['pSI_356_v2']–=–QUEEN\n",
      "                     (record='158431',–dbtype='addgene',–product='pSI_356_v2')\"\n",
      "                     /building_history_465=\"\"\n",
      "                     /building_history_465=\"pSI_827-ELG4OEK46KP17LQKNG13FABI\"\n",
      "                     /building_history_470=\"QUEEN.queried_features_dict['pSI_356\n",
      "                     _v2_f0']–=–QUEEN.dna_dict['pSI_356_v2'].searchsequence(quer\n",
      "                     y=cs.lib['BsmBI'].cutsite)\"\n",
      "                     /building_history_470=\"query: cs.lib['BsmBI'].cutsite\"\n",
      "                     /building_history_470=\"pSI_827-4R1R4ECVAKQ52S7EFXHRRQ94\"\n",
      "                     /building_history_471=\"QUEEN.dna_dict['pSI_356_v2_0']–=–cro\n",
      "                     pdna(QUEEN.dna_dict['pSI_356_v2'],–start=QUEEN.queried_feat\n",
      "                     ures_dict['pSI_356_v2_f0'][0],–end=QUEEN.queried_features_d\n",
      "                     ict['pSI_356_v2_f0'][1],–process_name='Digest–backbone',–pr\n",
      "                     ocess_description='Vector–backbone–plasmid–was–digested–wit\n",
      "                     h–BsmBI–(NEB–#R0739L)–or–BsaI–(NEB–#R3733L)–(see–Table–S2–f\n",
      "                     or–details)–during–the–Golden–Gate–Assembly.')\"\n",
      "                     /building_history_471=\"start: '3645/3649'; end:\n",
      "                     '1760/1764'\"\n",
      "                     /building_history_471=\"pSI_827-41UP8PKXVY149ZCTY190BVXJ\"\n",
      "                     /building_history_866=\"QUEEN.dna_dict['pSI_827']–=–joindna(\n",
      "                     *[QUEEN.dna_dict['dna_102'],–QUEEN.dna_dict['pSI_356_v2_0']\n",
      "                     ],–topology='circular',–product='pSI_827',–process_name='As\n",
      "                     semble',–process_description='Insert–and–vector–backbone–fr\n",
      "                     agments–were–mixed–together–with–BsmBI–(NEB–#R0739L)–or–Bsa\n",
      "                     I–(NEB–#R3733L)–(see–Table–S2–for–details)–according–to–the\n",
      "                     –manufacturer's–high–efficiency–Golden–Gate–Assembly–protoc\n",
      "                     ol.')\"\n",
      "                     /building_history_866=\"topology: circular\"\n",
      "                     /building_history_866=\"pSI_827-3BFM82O2245M3HL4K6IMALSR\"\n",
      "ORIGIN\n",
      "        1 aactatcgtc ttgagtccaa cccggtaaga cacgacttat cgccactggc agcagccact\n",
      "       61 ggtaacagga ttagcagagc gaggtatgta ggcggtgcta cagagttctt gaagtggtgg\n",
      "      121 cctaactacg gctacactag aagaacagta tttggtatct gcgctctgct gaagccagtt\n",
      "      181 accttcggaa aaagagttgg tagctcttga tccggcaaac aaaccaccgc tggtagcggt\n",
      "      241 ggtttttttg tttgcaagca gcagattacg cgcagaaaaa aaggatctca agaagatcct\n",
      "      301 ttgatctttt ctacggggtc tgacgctcag tggaacgaaa actcacgtta agggattttg\n",
      "      361 gtcatgagat tatcaaaaag gatcttcacc tagatccttt taaattaaaa atgaagtttt\n",
      "      421 aaatcaatct aaagtatata tgagtaaact tggtctgaca gttaccaatg cttaatcagt\n",
      "      481 gaggcaccta tctcagcgat ctgtctattt cgttcatcca tagttgcctg actccccgtc\n",
      "      541 gtgtagataa ctacgatacg ggagggctta ccatctggcc ccagtgctgc aatgataccg\n",
      "      601 cgagacccac gctcaccggc tccagattta tcagcaataa accagccagc cggaagggcc\n",
      "      661 gagcgcagaa gtggtcctgc aactttatcc gcctccatcc agtctattaa ttgttgccgg\n",
      "      721 gaagctagag taagtagttc gccagttaat agtttgcgca acgttgttgc cattgctaca\n",
      "      781 ggcatcgtgg tgtcacgctc gtcgtttggt atggcttcat tcagctccgg ttcccaacga\n",
      "      841 tcaaggcgag ttacatgatc ccccatgttg tgcaaaaaag cggttagctc cttcggtcct\n",
      "      901 ccgatcgttg tcagaagtaa gttggccgca gtgttatcac tcatggttat ggcagcactg\n",
      "      961 cataattctc ttactgtcat gccatccgta agatgctttt ctgtgactgg tgagtactca\n",
      "     1021 accaagtcat tctgagaata gtgtatgcgg cgaccgagtt gctcttgccc ggcgtcaata\n",
      "     1081 cgggataata ccgcgccaca tagcagaact ttaaaagtgc tcatcattgg aaaacgttct\n",
      "     1141 tcggggcgaa aactctcaag gatcttaccg ctgttgagat ccagttcgat gtaacccact\n",
      "     1201 cgtgcaccca actgatcttc agcatctttt actttcacca gcgtttctgg gtgagcaaaa\n",
      "     1261 acaggaaggc aaaatgccgc aaaaaaggga ataagggcga cacggaaatg ttgaatactc\n",
      "     1321 atactcttcc tttttcaata ttattgaagc atttatcagg gttattgtct catgagcgga\n",
      "     1381 tacatatttg aatgtattta gaaaaataaa caaatagggg ttccgcgcac atttccccga\n",
      "     1441 aaagtgccac ctgacgtcta agaaaccatt attatcatga cattaaccta taaaaatagg\n",
      "     1501 cgtatcacga ggccctttac cggttttccc atgattcctt catatttgca tatacgatac\n",
      "     1561 aaggctgtta gagagataat tagaattaat ttgactgtaa acacaaagat attagtacaa\n",
      "     1621 aatacgtgac gtagaaagta ataatttctt gggtagtttg cagttttaaa attatgtttt\n",
      "     1681 aaaatggact atcatatgct taccgtaact tgaaagtatt tcgatttctt ggctttatat\n",
      "     1741 atcttgtgga aaggacgaaa caccgcctta tgaccctgac acgctgtttt agagctagaa\n",
      "     1801 atagcaagtt aaaataaggc tagtccgtta tcaacttgaa aaagtggcac cgagtcggtg\n",
      "     1861 cttttttgtc acagcttgtc tgtaagcgga tgccgggagc agacaagccc gtcagggcgc\n",
      "     1921 gtcagcgggt gttggcgggt gtcggggctg gcttaactat gcggcatcag agcagattgt\n",
      "     1981 actgagagtg caccatatgc ggtgtgaaat accgcacaga tgcgtaagga gaaaataccg\n",
      "     2041 catcaggcgc cattcgccat tcaggctgcg caactgttgg gaagggcgat cggtgcgggc\n",
      "     2101 ctcttcgcta ttacgccagc tggcgaaagg gggatgtgct gcaaggcgat taagttgggt\n",
      "     2161 aacgccaggg ttttcccagt cacgacgttg taaaacgacg gccagtgaat tcgagctcgg\n",
      "     2221 tacccgggga tcctctagag tcgacctgca ggcatgcaag cttggcgtaa tcatggtcat\n",
      "     2281 agctgtttcc tgtgtgaaat tgttatccgc tcacaattcc acacaacata cgagccggaa\n",
      "     2341 gcataaagtg taaagcctgg ggtgcctaat gagtgagcta actcacatta attgcgttgc\n",
      "     2401 gctcactgcc cgctttccag tcgggaaacc tgtcgtgcca gctgcattaa tgaatcggcc\n",
      "     2461 aacgcgcggg gagaggcggt ttgcgtattg ggcgctcttc cgcttcctcg ctcactgact\n",
      "     2521 cgctgcgctc ggtcgttcgg ctgcggcgag cggtatcagc tcactcaaag gcggtaatac\n",
      "     2581 ggttatccac agaatcaggg gataacgcag gaaagaacat gtgagcaaaa ggccagcaaa\n",
      "     2641 aggccaggaa ccgtaaaaag gccgcgttgc tggcgttttt ccataggctc cgcccccctg\n",
      "     2701 acgagcatca caaaaatcga cgctcaagtc agaggtggcg aaacccgaca ggactataaa\n",
      "     2761 gataccaggc gtttccccct ggaagctccc tcgtgcgctc tcctgttccg accctgccgc\n",
      "     2821 ttaccggata cctgtccgcc tttctccctt cgggaagcgt ggcgctttct catagctcac\n",
      "     2881 gctgtaggta tctcagttcg gtgtaggtcg ttcgctccaa gctgggctgt gtgcacgaac\n",
      "     2941 cccccgttca gcccgaccgc tgcgccttat ccggt\n",
      "//\n"
     ]
    }
   ],
   "source": [
    "#Step 3: Assemble\n",
    " \n",
    "assembled_plasmids = []\n",
    "i = 0\n",
    "while i < len(golden_gate_assembly_2):\n",
    "    assembly = joindna(annealed_DNA[i],\n",
    "                       digested_backbones[i],\n",
    "                       product = golden_gate_assembly_2['plasmid_id'][i],\n",
    "                       topology = \"circular\", pn=pn3, pd=pd3)\n",
    "    assembled_plasmids.append(assembly)\n",
    "    i = i+1\n",
    "\n",
    "assembled_plasmids[0].outputgbk()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49fa9700-3e36-45d8-984f-b53355206937",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step 4: Output QUEEN gbk file\n",
    "\n",
    "i = 0\n",
    "while i < len(golden_gate_assembly_2):\n",
    "    assembled_plasmids[i].outputgbk(output = golden_gate_assembly_2['plasmid_id'][i]+ '.gbk')\n",
    "    i = i+1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2410,
   "id": "f54b11ab-ef91-45f9-b9fd-f840bc75a2ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QUEEN gbk files repaired.\n"
     ]
    }
   ],
   "source": [
    "#Step 5: Repair gbk locus ID (temporary fix; will be implemented in next update of QUEEN)\n",
    "\n",
    "i = 0\n",
    "while i < len(golden_gate_assembly_2):\n",
    "    gbk_file = './' + golden_gate_assembly_2['plasmid_id'][i]+ '.gbk'\n",
    "    handle = open(gbk_file, 'r')\n",
    "    record = GenBank.read(handle)\n",
    "    record.locus = golden_gate_assembly_2['plasmid_id'][i]\n",
    "    record.accession = ''\n",
    "    record.version = ''\n",
    "    record.definition = 'Plasmid was generated by QUEEN v1.2.0 (https://github.com/yachielab/QUEEN)'\n",
    "    record.organism = 'Synthetic DNA construct'\n",
    "    fout = open(gbk_file, 'w') # final gbk file name\n",
    "    fout.write(str(record))\n",
    "    fout.close()\n",
    "    i = i+1\n",
    "    \n",
    "print('QUEEN gbk files repaired.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
